File: norma/src/ingest.py
# ingest.py — builds FAISS index + metadata from preprocessor chunks.json
import argparse
import json
import os
from pathlib import Path

import faiss
import numpy as np
from sentence_transformers import SentenceTransformer


DEFAULT_MODEL = "sentence-transformers/all-MiniLM-L6-v2"


def load_chunks(path: Path):
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, list):
        raise ValueError("chunks.json must be a list of chunk objects.")
    return data


def build_embeddings(chunks, model, batch_size=32):
    texts = [c.get("text", "") for c in chunks]
    embeddings = model.encode(texts, batch_size=batch_size, show_progress_bar=True, convert_to_numpy=True)
    embeddings = embeddings.astype("float32")
    return embeddings


def build_faiss_index(embeddings):
    dim = embeddings.shape[1]
    faiss.normalize_L2(embeddings)
    index = faiss.IndexFlatIP(dim)  # cosine via normalized vectors + inner product
    index.add(embeddings)
    return index


def create_metadata(chunks):
    metadata = []
    for i, c in enumerate(chunks):
        meta = {
            "chunk_id": i,
            # unify name as "page" for app.py
            "page": c.get("start_page") or c.get("page"),
            "section": c.get("section") or c.get("section_hint", "")
        }
        metadata.append(meta)
    return metadata


def save_all(index, embeddings, metadata, out_index: Path, out_emb: Path, out_meta: Path):
    out_index.parent.mkdir(parents=True, exist_ok=True)
    faiss.write_index(index, str(out_index))
    np.save(str(out_emb), embeddings)
    with open(out_meta, "w", encoding="utf-8") as f:
        json.dump(metadata, f, indent=2, ensure_ascii=False)
    print(f"Saved index -> {out_index}")
    print(f"Saved embeddings -> {out_emb}")
    print(f"Saved metadata -> {out_meta}")


def parse_args():
    p = argparse.ArgumentParser(description="Build FAISS index from chunks.json")
    p.add_argument("--chunks", type=Path, required=False,
                   default=Path(os.getenv("NORMA_CHUNKS_PATH", "D:/Projects/Specora/Norma/data/processed/chunks.json")))
    p.add_argument("--index", type=Path, required=False,
                   default=Path(os.getenv("NORMA_INDEX_PATH", "D:/Projects/Specora/Norma/data/processed/faiss_index.bin")))
    p.add_argument("--meta", type=Path, required=False,
                   default=Path(os.getenv("NORMA_META_PATH", "D:/Projects/Specora/Norma/data/processed/chunk_metadata.json")))
    p.add_argument("--emb", type=Path, required=False,
                   default=Path(os.getenv("NORMA_EMB_PATH", "D:/Projects/Specora/Norma/data/processed/embeddings.npy")))
    p.add_argument("--model", type=str, default=os.getenv("NORMA_EMBED_MODEL", DEFAULT_MODEL))
    return p.parse_args()


def main():
    args = parse_args()

    print("Loading chunks from:", args.chunks)
    chunks = load_chunks(args.chunks)

    print("Loading embedding model:", args.model)
    model = SentenceTransformer(args.model)

    print("Building embeddings...")
    embeddings = build_embeddings(chunks, model)

    print("Building FAISS index...")
    index = build_faiss_index(embeddings)

    print("Creating metadata...")
    metadata = create_metadata(chunks)

    print("Saving artifacts...")
    save_all(index, embeddings, metadata, args.index, args.emb, args.meta)

    print("Ingestion complete.")


if __name__ == "__main__":
    main()

File: be/scripts/ingest.js
#!/usr/bin/env node

/**
 * ingest.js — Builds FAISS index + metadata from chunks.json.
 * Node.js port of norma/src/ingest.py
 * 
 * Usage:
 *   node scripts/ingest.js
 *   node scripts/ingest.js --chunks ./data/norma/processed/chunks.json --index ./data/norma/processed/faiss_index.bin --meta ./data/norma/processed/chunk_metadata.json
 */

import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { resolve, dirname } from "path";
import faiss from "faiss-node";
import { EmbeddingModel } from "fastembed";
import dotenv from "dotenv";

const { IndexFlatIP } = faiss;

dotenv.config();

// -----------------------
// Config
// -----------------------
const args = process.argv.slice(2);
function getArg(name, fallback) {
    const idx = args.indexOf(name);
    return idx !== -1 && args[idx + 1] ? args[idx + 1] : fallback;
}

const CHUNKS_PATH = resolve(getArg("--chunks", process.env.NORMA_CHUNKS_PATH || "./data/norma/processed/chunks.json"));
const INDEX_PATH = resolve(getArg("--index", process.env.NORMA_INDEX_PATH || "./data/norma/processed/faiss_index.bin"));
const META_PATH = resolve(getArg("--meta", process.env.NORMA_META_PATH || "./data/norma/processed/chunk_metadata.json"));
const MODEL_NAME = getArg("--model", process.env.NORMA_EMBED_MODEL || "sentence-transformers/all-MiniLM-L6-v2");

// -----------------------
// Helpers
// -----------------------

/**
 * L2-normalize a Float32Array vector in-place.
 */
function normalizeL2(vec) {
    let norm = 0;
    for (let i = 0; i < vec.length; i++) {
        norm += vec[i] * vec[i];
    }
    norm = Math.sqrt(norm);
    if (norm > 0) {
        for (let i = 0; i < vec.length; i++) {
            vec[i] /= norm;
        }
    }
    return vec;
}

// -----------------------
// Main
// -----------------------
async function main() {
    // Load chunks
    console.log("Loading chunks from:", CHUNKS_PATH);
    const chunks = JSON.parse(readFileSync(CHUNKS_PATH, "utf-8"));
    if (!Array.isArray(chunks)) {
        throw new Error("chunks.json must be a list of chunk objects.");
    }
    console.log(`Loaded ${chunks.length} chunks.`);

    // Load embedding model
    console.log("Loading embedding model:", MODEL_NAME);
    const model = await EmbeddingModel.from(MODEL_NAME);

    // Build embeddings
    console.log("Building embeddings...");
    const texts = chunks.map((c) => c.text || "");
    const allEmbeddings = [];

    for await (const batch of model.embed(texts)) {
        allEmbeddings.push(batch);
    }

    // Flatten all batches into a single array of Float32Arrays
    // fastembed returns batches — each batch is a Float32Array for one text
    const embeddings = allEmbeddings;

    if (embeddings.length === 0) {
        throw new Error("No embeddings were generated.");
    }

    const dim = embeddings[0].length;
    console.log(`Generated ${embeddings.length} embeddings of dimension ${dim}.`);

    // Normalize all embeddings
    console.log("Normalizing embeddings...");
    for (const emb of embeddings) {
        normalizeL2(emb);
    }

    // Build FAISS index (IndexFlatIP for cosine via normalized vectors + inner product)
    console.log("Building FAISS index...");
    const index = new IndexFlatIP(dim);

    for (const emb of embeddings) {
        index.add(emb);
    }

    console.log(`FAISS index built with ${index.ntotal()} vectors.`);

    // Create metadata
    console.log("Creating metadata...");
    const metadata = chunks.map((c, i) => ({
        chunk_id: i,
        page: c.start_page || c.page || null,
        section: c.section || c.section_hint || "",
    }));

    // Save artifacts
    mkdirSync(dirname(INDEX_PATH), { recursive: true });

    index.write(INDEX_PATH);
    console.log(`Saved index -> ${INDEX_PATH}`);

    writeFileSync(META_PATH, JSON.stringify(metadata, null, 2), "utf-8");
    console.log(`Saved metadata -> ${META_PATH}`);

    console.log("Ingestion complete.");
}

main().catch((err) => {
    console.error("Ingestion failed:", err);
    process.exit(1);
});
